teps/0138-decouple-api-and-feature-versioning.md

```diff
@@ -72,8 +91,331 @@ Per [TEP-0033](https://github.com/tektoncd/community/blob/main/teps/0033-tekton-
+
+
+## Additional CI tests
+### Additional Test Combinations
+We cannot test 2\*\*N combinations of per feature flags since that would be too time consuming.
+Therefore, we try to mimic the following scenarios.
+
+- Say, a cluster operator has set the cluster to run Stable features. Occasionally, they may want to turn on a feature that is not yet stable. 
+- Similarly if the cluster is running Beta features by default (i.e. all feature flags at a beta stability level are ON and enable-api-fields is set to “stable”), cluster operators may want to turn on an individual feature at an alpha level. 
+- Conversely, if the cluster is running all the features (all feature flags are ON and enable api fields is set to “alpha”), they may want to turn off individual features.
+
+
+
+||enable-api-fields|Per feature flags|Integration tests|
+| :- | :- | :- | :- |
+|Opt-in stable|Set to stable|<p>All feature flags are OFF by default.</p><p>Turn ON one feature flag at a time. </p>|Run a small number of e2e tests against N combinations. It is not feasible to run the entire e2e test suite against N combinations.|
+|Opt-in beta|Set to beta|Turn ON all beta per-feature flags by default.<br><br>Turn ON one per-feature (flag at an alpha stability level) at a time.|Run a small number of e2e tests against M combinations (where M <= N)|
+|Opt-in alpha|Set to alpha|Turn ON all per-feature flags by default.<br><br>TURN OFF one feature flag at a time.|Run a small number of e2e tests against N combinations.|
+
+### <a name="_tcstxie74non"></a>How many tests can we run in a reasonable amount of time?
+Based on a [recent PR](https://github.com/tektoncd/pipeline/pull/7032), our integration tests take between 26 mins (stable) → 33 mins (alpha). We don’t want to go beyond that. Based on [Feature flags testing matrix](https://docs.google.com/document/d/1r_MX9-mzRtdbfNQq5VC4guHb-tphA0WxWhlQIsusEEA/edit?resourcekey=0-RALry7-GaKn9i19UEaRnYg) benchmarking, the approximate time is:
+
+T = N<sub>pipelines</sub>\*N<sub>tasks</sub>\*N<sub>features</sub>\*6 s
+
+
+
+Assume: 
+
+N<sub>features</sub> = 20 (currently we have 17 api features; lets assume that at any point in time, we will likely have ~20 individual features)
+N<sub>tasks</sub> = 2 (two tasks per pipeline)
+
+
+|Scenario|N<sub>pipelines</sub>|N<sub>features</sub>|N<sub>tasks/pipeline</sub>|T (mins)|
+| :- | :- | :- | :- | :- |
+|How many pipelines can we afford to run in 30 mins?|**7**|20|2|30|
+|How long would it take to run a single (same) pipeline for all the features?|1|20|2|**4**|
+
+### How many tests should we run against the additional tests?
+- It is not feasible to run the entire e2e test suite against N features.
+- Coming up with a single test that covers all common core features might be challenging.
+  - We probably want our tests to be able to cover
+    - params
+    - workspaces
+    - results
+    - DAG (small fan out)?
+    - Finally
+    - remote resolution
```

Comment:
The list feels long but we should follow the practice of testing a real world use cases which involves mostly all of the features in a single pipeline.